packages/graphql-client/src/graphql-client.ts

```diff
@@ -36,32 +48,93 @@ async function processJSONResponse<TData = any>(
   };
 }
 
+function validateRetries(retries: number | undefined) {
+  if (retries !== undefined && retries < 0) {
+    throw new Error(
+      `${ERROR_PREFIX}: The provided "retries" value (${retries}) is invalid - it cannot be less than zero.`
+    );
+  }
+}
+
 export function createGraphQLClient<TClientOptions extends ClientOptions>({
   headers,
   url,
   fetchAPI = fetch,
+  retries = 0,
 }: TClientOptions): GraphQLClient {
+  validateRetries(retries);
+
   const config = {
     headers,
     url,
+    retries,
   };
 
-  const fetch: GraphQLClient["fetch"] = (operation, options = {}) => {
-    const { variables, headers: overrideHeaders, url: overrideUrl } = options;
+  const httpFetch = async (
+    params: Parameters<CustomFetchAPI>,
+    count: number,
+    maxTries: number
+  ): ReturnType<GraphQLClient["fetch"]> => {
+    const nextCount = count + 1;
+    let resetTime = RETRY_WAIT_TIME;
+    try {
+      const response = await fetchAPI(...params);
+      if (
+        !response.ok &&
+        RETRIABLE_STATUS_CODES.includes(response.status) &&
+        nextCount <= maxTries
+      ) {
+        const retryAfter = response.headers.get("Retry-After");
+        if (typeof retryAfter !== "undefined" && retryAfter !== null) {
+          resetTime = parseFloat(retryAfter) * 1000;
+        }
+        throw new Error();
+      }
+
+      return response;
+    } catch (error) {
+      if (nextCount <= maxTries) {
+        await sleep(resetTime);
+        return httpFetch(params, nextCount, maxTries);
+      }
+
```

Comment:
Note: This error only occurs when the request was abandoned (eg. no network). 

If the server has responded, the client will return the last server response (as per the previous block).